master/buildbot/process/build.py

```diff
@@ -396,110 +404,118 @@ def startBuild(self, workerforbuilder: AbstractWorkerForBuilder):
         yield self.master.data.updates.setBuildStateString(self.buildid, 'starting')
         yield self.master.data.updates.generateNewBuildEvent(self.buildid)
 
-        try:
-            self.setupBuild()  # create .steps
-        except Exception:
-            yield self.buildPreparationFailure(Failure(), "setupBuild")
-            yield self.buildFinished(['Build.setupBuild', 'failed'], EXCEPTION)
-            return
+        if not self.skipBuild:
```

Comment:
I think it makes sense to have the following here:

```
if self.skipBuild:
   ...
   return
```

This short circuit will make the rest of the code unchanged and easier to understand. We get a little bit of duplication (10 lines?), but this code path is so different that the usual reasons to eliminate duplication don't apply anymore.

Reply:
Maybe it makes sense to do this just before `self._preparation_step` is setup, for the same reasons.